src/tactic/itauto.lean

```diff
@@ -0,0 +1,486 @@
+| (proof.or_elim _ _ p x q r) :=
+  format!"({p.to_format}.elim (λ {x}, {q.to_format}) (λ {x}, {r.to_format})"
+| (proof.imp_imp_simp _ _ _ p) := format!"(imp_imp_simp {p.to_format})"
+
+meta instance : has_to_format proof := ⟨proof.to_format⟩
+
+/-- A variant on `proof.exfalso'` that performs opportunistic simplification. -/
+meta def proof.exfalso : prop → proof → proof
+| prop.false p := p
+| A p := proof.exfalso' A p
+
+/-- A variant on `proof.app'` that performs opportunistic simplification.
+(This doesn't do full normalization because we don't want the proof size to blow up.) -/
+meta def proof.app : proof → proof → proof
+| (proof.curry ak _ B p) q := proof.curry₂ ak B p q
+| (proof.curry₂ ak _ p q) r := p.app (q.and_intro ak r)
+| (proof.or_imp_left _ B p) q := p.app (q.or_inl B)
+| (proof.or_imp_right A _ p) q := p.app (q.or_inr A)
+| (proof.imp_imp_simp x A _ p) q := p.app (proof.intro x A q)
+| p q := p.app' q
+
+/-- A typechecker for the `proof` type. This is not used by the tactic but can be used for
+debugging. -/
+meta def proof.check : name_map prop → proof → option prop
+| Γ (proof.hyp i) := Γ.find i
+| Γ proof.triv := some prop.true
+| Γ (proof.exfalso' A p) := guard (p.check Γ = some prop.false) $> A
+| Γ (proof.intro x A p) := do B ← p.check (Γ.insert x A), pure (prop.imp A B)
+| Γ (proof.and_left ak p) := do
+  prop.and ak' A B ← p.check Γ | none,
+  guard (ak = ak') $> (ak.sides A B).1
+| Γ (proof.and_right ak p) := do
+  prop.and ak' A B ← p.check Γ | none,
+  guard (ak = ak') $> (ak.sides A B).2
+| Γ (proof.and_intro and_kind.and p q) := do
+  A ← p.check Γ, B ← q.check Γ,
+  pure (A.and and_kind.and B)
+| Γ (proof.and_intro ak p q) := do
+  prop.imp A B ← p.check Γ | none,
+  C ← q.check Γ, guard (C = prop.imp B A) $> (A.and ak B)
+| Γ (proof.curry ak A' B' p) := do
+  prop.imp (prop.and ak' A B) C ← p.check Γ | none,
+  guard (ak = ak' ∧ ak.sides A B = (A', B')) $> (A'.imp $ B'.imp C)
+| Γ (proof.curry₂ ak B' p q) := do
+  prop.imp (prop.and ak' A B) C ← p.check Γ | none,
+  A' ← q.check Γ, guard (ak = ak' ∧ ak.sides A B = (A', B')) $> (B'.imp C)
+| Γ (proof.app' p q) := do prop.imp A B ← p.check Γ | none, A' ← q.check Γ, guard (A = A') $> B
+| Γ (proof.or_imp_left A B p) := do
+  prop.imp (prop.or A' B') C ← p.check Γ | none,
+  guard (A = A' ∧ B = B') $> (A.imp C)
+| Γ (proof.or_imp_right A B p) := do
+  prop.imp (prop.or A' B') C ← p.check Γ | none,
+  guard (A = A' ∧ B = B') $> (B.imp C)
+| Γ (proof.or_inl B p) := do A ← p.check Γ | none, pure (A.or B)
+| Γ (proof.or_inr A p) := do B ← p.check Γ | none, pure (A.or B)
+| Γ (proof.or_elim A' B' p x q r) := do
+  prop.or A B ← p.check Γ | none,
+  C ← q.check (Γ.insert x A),
+  C' ← r.check (Γ.insert x B),
+  guard (A = A' ∧ B = B' ∧ C = C') $> C
+| Γ (proof.imp_imp_simp x A B p) := do
+  prop.imp (prop.imp A' B') C ← p.check Γ | none,
+  guard (A = A' ∧ B = B') $> (B.imp C)
+
+/-- Get a new name in the pattern `h0, h1, h2, ...` -/
```

Comment:
What happens in the following?
```lean
example (h0 : q) : p → q := by itauto
```

Reply:
The names in the context are added via their unique names, so they don't collide with this naming system and the prover itself doesn't confuse them. When the proof is complete, these `h0` names are used for the pp names in the binders, but lean is smart enough to disambiguate in printing:
```lean
theorem T (p q : Prop) (h0 : q) : p → q := by itauto
#print T
-- theorem T : ∀ (p q : Prop), q → p → q :=
-- λ (p q : Prop) (h0 : q) (h0_1 : p), h0
```